Hud overhaul

kaadmy opened the pull request:
Currently this basically makes the HUD more sane and adds a notification feed for obituaries.

- Shows the last few lines of chat history when you open the chat prompt
- Put player connect/disconnect/timeout/name change/team change messages in chat
- Add a notification feed for obituaries and eventually stuff like flag drops/captures/returns/takes
- Moved health and armor to the middle of the screen
- Moved ammo to the bottom-left corner of the screen
- Fixes the bug that allowed you to see player names through walls

Please don't merge yet; this will probably only compile on Linux and needs additional work to get Mac and Windows build files working.

**Edit:** We'll need icons for the obituary/notification feed eventually before this is merged.

Paril commented on src/cgame/default/cg_hud.c:
```diff
@@ -1263,6 +1552,19 @@ void Cg_InitHud(void) {
+ * @brief
+ */
+void Cg_ShutdownHud(void) {
+	g_slist_free(notifications.items);
```
Doesn't free bg_notification_item_t instances. Use g_slist_free_full with g_free0 as second argument.

kaadmy replied:
I don't think `g_free0` exists..?

Paril replied:
g_free, whatever. I don't even know why they called it g_malloc0.

jdolan replied:
Because it returns zero'd memory.

Paril commented on src/cgame/default/cg_hud.c:
```diff
@@ -1175,6 +1233,231 @@ static void Cg_DrawTargetName(const player_state_t *ps) {
+
+				x -= NOTIFICATION_PIC_HEIGHT + NOTIFICATION_PADDING_X;
+
+				cgi.DrawImage(x, y, ((vec_t) NOTIFICATION_PIC_HEIGHT) / HUD_PIC_HEIGHT, cgi.LoadImage(item->pic, IT_PIC));
```
Precache the pic in the parse function and use that, don't keep the raw string for the pic. It's immutable, so there's no sense in having the whole string be carried along.

Paril commented on src/cgame/default/cg_hud.c:
```diff
@@ -1175,6 +1233,231 @@ static void Cg_DrawTargetName(const player_state_t *ps) {
+
+				x -= NOTIFICATION_PIC_HEIGHT + NOTIFICATION_PADDING_X;
+
+				cgi.DrawImage(x, y, ((vec_t) NOTIFICATION_PIC_HEIGHT) / HUD_PIC_HEIGHT, cgi.LoadImage(item->pic, IT_PIC));
```
Ditto, see below.

Paril commented on src/cgame/default/cg_hud.c:
```diff
@@ -1175,6 +1233,231 @@ static void Cg_DrawTargetName(const player_state_t *ps) {
+				x -= NOTIFICATION_PIC_HEIGHT + NOTIFICATION_PADDING_X;
+
+				cgi.DrawImage(x, y, ((vec_t) NOTIFICATION_PIC_HEIGHT) / HUD_PIC_HEIGHT,
+					cgi.LoadImage(Bg_GetModIcon(item->mod, item->mod & MOD_FRIENDLY_FIRE), IT_PIC));
```
Ditto, see below.

Paril commented on src/cgame/default/cg_hud.c:
```diff
@@ -1175,6 +1233,231 @@ static void Cg_DrawTargetName(const player_state_t *ps) {
+				x -= NOTIFICATION_PIC_HEIGHT + NOTIFICATION_PADDING_X;
+
+				cgi.DrawImage(x, y, ((vec_t) NOTIFICATION_PIC_HEIGHT) / HUD_PIC_HEIGHT,
+					cgi.LoadImage(Bg_GetModIcon(item->mod, item->mod & MOD_FRIENDLY_FIRE), IT_PIC));
```
Ditto, see below.

Paril commented on src/cgame/default/cg_hud.c:
```diff
@@ -1175,6 +1233,231 @@ static void Cg_DrawTargetName(const player_state_t *ps) {
+			break;
+		case NOTIFICATION_TYPE_OBITUARY_PIC: {
+			const char *p = cgi.ReadString();
+			strcpy(item->pic, p);
```
"pic" should be cgi.LoadImage'd. I'm not certain, but if the obituary pictures are entirely constant (that is, the pics you send are *always* consistent between game and cgame for obituaries), send the index to the array and not the pic itself. Then we can reference a cgame-specific array of the images that are precached in this file.

kaadmy replied:
Oh yeah, I should.

Paril commented on src/cgame/default/cg_hud.c:
```diff
@@ -1175,6 +1233,231 @@ static void Cg_DrawTargetName(const player_state_t *ps) {
+			break;
+		case NOTIFICATION_TYPE_PLAYER_ACTION: {
+			const char *p = cgi.ReadString();
+			strcpy(item->pic, p);
```
Ditto, cgi.LoadImage here.

Paril commented on src/game/default/bg_notification.h:
```diff
@@ -0,0 +1,58 @@
+	uint16_t client_id_1; // CID
+	uint16_t client_id_2; // CID
+
+	char pic[MAX_QPATH]; // PIC
```
should be uint16_t for a pic index.

Paril commented on src/game/default/g_types.h:
```diff
@@ -355,6 +357,42 @@ typedef enum {
+#define MOD_HOOK					29
+#define MOD_ACT_OF_GOD				30
+#define MOD_FRIENDLY_FIRE			0x80
+
```
Why dis not enum

kaadmy replied:
It is an enum in the weapon_tweaks branch, I was going to merge that first since I added a new MOD too.

Paril replied:
Since this PR ends up modifying MODs, I'd do it here and then cherry pick it to the other branch when this one merges into master.

kaadmy replied:
Well the other PR also changes MODs.

Paril commented on src/cgame/default/cg_hud.c:
```diff
@@ -136,40 +151,56 @@ static void Cg_DrawVital(r_pixel_t x, const int16_t value, const int16_t icon, i
+		Cg_DrawVital(x, true, health, health_icon, HUD_HEALTH_MED, HUD_HEALTH_LOW);
 	}
 
 	if (atoi(cgi.ConfigString(CS_GAMEPLAY)) != GAME_INSTAGIB) {
```
One thing I've been meaning to do but haven't gotten around to is finding a common place in cgame to store the results of these cgi.ConfigString fetch operations. Seems silly to parse this every single frame multiple times.

kaadmy replied:
Should I try to cache config strings for this PR?

jdolan replied:
Good luck finding config string parsing in a profiler ;) Hint: it's nothing.

kaadmy replied:
...why in a profiler?

jdolan replied:
I'm saying, config string parsing would never show up on a profiler. Don't worry about it.

Paril commented on src/cgame/default/cg_hud.c:
```diff
@@ -1107,12 +1144,29 @@ static void Cg_DrawSelectWeapon(const player_state_t *ps) {
+		cg_select_weapon_fade->modified = false;
+
+		cg_select_weapon_fade->value = Clamp(cg_select_weapon_fade->value, 0.0, cg_select_weapon_interval->value);
+	}
```
Fade is a good idea for this, I need to test it to see how it feels though. If I was to implement fade it would be an additional amount of seconds added to the display interval probably.

jdolan replied:
What does this do? Blend the weapon during the switch animation?

kaadmy replied:
No, it fades out the weapon bar a second or so after switching weapons.

kaadmy commented:
On mobile right now, but everything I either already know or was going to
change :P

The MOD is an enum in the weapon_tweaks branch, I wanted to merge that
first.

The millis/sec stuff I kept as millis because it already was.


I'm not very surprised about the memory stuff, I'll go through it tomorrow.

On Jun 6, 2017 9:08 PM, "Jonathan" <notifications@github.com> wrote:

*@Paril* requested changes on this pull request.
------------------------------

In src/game/default/g_client_stats.c
<https://github.com/jdolan/quetoo/pull/487#discussion_r120529643>:

> @@ -270,6 +270,12 @@ void G_ClientStats(g_entity_t *ent) {
 		client->ps.stats[STAT_READY] = client->locals.persistent.ready;
 	}

+	// respawn time
+	client->ps.stats[STAT_RESPAWN] = 0;
+	if (g_level.time < client->locals.respawn_time) {
+		client->ps.stats[STAT_RESPAWN] = client->locals.respawn_time - g_level.time;

For efficiency-sake, and since the client knows about server time (I'm
pretty sure it does, at least, since it knows frames), maybe you should
just pass over the raw respawn time and let the client calculate the time
left themselves? Then the stats structure isn't changing every single frame.
------------------------------

In src/cgame/default/cg_hud.c
<https://github.com/jdolan/quetoo/pull/487#discussion_r120530040>:

> +}
+
+/**
+ * @brief
+ */
+static void Cg_DrawNotification(const player_state_t *ps) {
+	r_pixel_t cw, ch, x, y;
+	bg_notification_item_t *item;
+
+	GSList *list;
+
+	for (int32_t i = 0; i < notifications.num_lines; i++) {
+		list = g_slist_nth(notifications.items, i);
+
+		if (list && cgi.client->unclamped_time > ((bg_notification_item_t
*) list->data)->when) {
+			notifications.items = g_slist_delete_link(notifications.items, list);

This doesn't free the actual underlying data, btw. I also recommend that
you use the proper method for iterating slists, because g_slist_nth is O(n)
and will be slower every time you call it. Just loop through it and store
the "next" GSlist, that way when you remove the current one you still have
a next one to go to. It's safe to do so. Also, make sure to g_free0 the
actual bg_notification_item_t, as you're leaking memory here.
------------------------------

In src/cgame/default/cg_hud.c
<https://github.com/jdolan/quetoo/pull/487#discussion_r120530128>:

> +			item->when = 8000;
+
+			if (!cg_draw_notifications_disable_print->value) {
+				cgi.Print("^7%s ^7%s\n", cgi.client->client_info[item->client_id_1].name,
+					item->string_1);
+			}
+		}
+			break;
+		default:
+			cgi.Warn("Invalid notification type %d\n", item->type);
+			break;
+	}
+
+	item->when = cgi.client->unclamped_time + (item->when *
cg_draw_notifications_time->value);
+
+	notifications.items = g_slist_prepend(notifications.items, item);

Might want to have an upper cap just so that large games don't completely
flood the right side of the screen.

EDIT: I see upper cap is already implemented for display as
cg_draw_notifications_lines, but I'm guessing you don't use that here
because some notifications have a longer display time. Maybe you could use
cg_draw_notifications_lines->integer * 2 as the upper cap?
------------------------------

In src/cgame/default/cg_hud.c
<https://github.com/jdolan/quetoo/pull/487#discussion_r120530207>:

> @@ -1263,6 +1552,19 @@ void Cg_InitHud(void) {
 	cgi.Cmd("cg_weapon_next", Cg_Weapon_Next_f, CMD_CGAME, "Open the
weapon bar to the next weapon. In chasecam, switches to next
target.");
 	cgi.Cmd("cg_weapon_previous", Cg_Weapon_Prev_f, CMD_CGAME, "Open the
weapon bar to the previous weapon. In chasecam, switches to previous
target.");

-	cg_select_weapon_delay = cgi.Cvar("cg_select_weapon_delay", "250",
CVAR_ARCHIVE, "The amount of time, in milliseconds, to wait between
changing weapons in the scroll view. Clicking will override this value
and switch immediately.");
-	cg_select_weapon_interval = cgi.Cvar("cg_select_weapon_interval",
"750", CVAR_ARCHIVE, "The amount of time, in milliseconds, to show the
weapon bar after changing weapons.");
+	cg_select_weapon_alpha = cgi.Cvar("cg_select_weapon_alpha", "0.5",
CVAR_ARCHIVE,
+					  "The opacity of unselected weapons in the weapon bar.");
+	cg_select_weapon_delay = cgi.Cvar("cg_select_weapon_delay", "250",
CVAR_ARCHIVE,
+					  "The amount of time, in milliseconds, to wait between changing
weapons in the scroll view. Clicking will override this value and
switch immediately.");
+	cg_select_weapon_fade = cgi.Cvar("cg_select_weapon_fade", "800", CVAR_ARCHIVE,
+					 "The amount of time, in milliseconds, for the weapon bar to
fade out in.");
+	cg_select_weapon_interval = cgi.Cvar("cg_select_weapon_interval",
"1600", CVAR_ARCHIVE,
+					     "The amount of time, in milliseconds, to show the weapon
bar after changing weapons.");

jdolan mentioned that we're towing a line here with having some cvars take
milliseconds and some take seconds. It might be worth fixing these in this
branch, since the millisecond ones were mostly stuff I had added (stains
and these). I think seconds are a better option.
------------------------------

In src/cgame/default/cg_hud.c
<https://github.com/jdolan/quetoo/pull/487#discussion_r120530251>:

> -	cg_select_weapon_interval = cgi.Cvar("cg_select_weapon_interval", "750", CVAR_ARCHIVE, "The amount of time, in milliseconds, to show the weapon bar after changing weapons.");
+	cg_select_weapon_alpha = cgi.Cvar("cg_select_weapon_alpha", "0.5",
CVAR_ARCHIVE,
+					  "The opacity of unselected weapons in the weapon bar.");
+	cg_select_weapon_delay = cgi.Cvar("cg_select_weapon_delay", "250",
CVAR_ARCHIVE,
+					  "The amount of time, in milliseconds, to wait between changing
weapons in the scroll view. Clicking will override this value and
switch immediately.");
+	cg_select_weapon_fade = cgi.Cvar("cg_select_weapon_fade", "800", CVAR_ARCHIVE,
+					 "The amount of time, in milliseconds, for the weapon bar to
fade out in.");
+	cg_select_weapon_interval = cgi.Cvar("cg_select_weapon_interval",
"1600", CVAR_ARCHIVE,
+					     "The amount of time, in milliseconds, to show the weapon
bar after changing weapons.");
+}
+
+/**
+ * @brief
+ */
+void Cg_ShutdownHud(void) {
+	g_slist_free(notifications.items);

Doesn't free bg_notification_item_t instances. Use g_slist_free_full with
g_free0 as second argument.
------------------------------

In src/cgame/default/cg_hud.c
<https://github.com/jdolan/quetoo/pull/487#discussion_r120530427>:

> +
+				name = cgi.client->client_info[item->client_id_1].name;
+
+				x -= cgi.StringWidth(name) + NOTIFICATION_PADDING_X;
+
+				cgi.DrawString(x, y + 6, name, CON_COLOR_DEFAULT);
+			}
+				break;
+			case NOTIFICATION_TYPE_PLAYER_ACTION: {
+				x -= cgi.StringWidth(item->string_1) + NOTIFICATION_PADDING_X;
+
+				cgi.DrawString(x, y + 6, item->string_1, CON_COLOR_DEFAULT);
+
+				x -= NOTIFICATION_PIC_HEIGHT + NOTIFICATION_PADDING_X;
+
+				cgi.DrawImage(x, y, ((vec_t) NOTIFICATION_PIC_HEIGHT) /
HUD_PIC_HEIGHT, cgi.LoadImage(item->pic, IT_PIC));

Precache the pic in the parse function and use that, don't keep the raw
string for the pic. It's immutable, so there's no sense in having the whole
string be carried along.
------------------------------

In src/cgame/default/cg_hud.c
<https://github.com/jdolan/quetoo/pull/487#discussion_r120530432>:

> +				x -= NOTIFICATION_PIC_HEIGHT + NOTIFICATION_PADDING_X;
+
+				cgi.DrawImage(x, y, ((vec_t) NOTIFICATION_PIC_HEIGHT) / HUD_PIC_HEIGHT,
+					cgi.LoadImage(Bg_GetModIcon(item->mod, item->mod &
MOD_FRIENDLY_FIRE), IT_PIC));
+			}
+				break;
+			case NOTIFICATION_TYPE_OBITUARY_PIC: {
+				char *name = cgi.client->client_info[item->client_id_2].name;
+
+				x -= cgi.StringWidth(name) + NOTIFICATION_PADDING_X;
+
+				cgi.DrawString(x, y + 6, name, CON_COLOR_DEFAULT);
+
+				x -= NOTIFICATION_PIC_HEIGHT + NOTIFICATION_PADDING_X;
+
+				cgi.DrawImage(x, y, ((vec_t) NOTIFICATION_PIC_HEIGHT) /
HUD_PIC_HEIGHT, cgi.LoadImage(item->pic, IT_PIC));

Ditto, see below.
------------------------------

In src/cgame/default/cg_hud.c
<https://github.com/jdolan/quetoo/pull/487#discussion_r120530441>:

> +				x -= cgi.StringWidth(name) + NOTIFICATION_PADDING_X;
+
+				cgi.DrawString(x, y + 6, name, CON_COLOR_DEFAULT);
+			}
+				break;
+			case NOTIFICATION_TYPE_OBITUARY_SELF: {
+				char *name = cgi.client->client_info[item->client_id_1].name;
+
+				x -= cgi.StringWidth(name) + NOTIFICATION_PADDING_X;
+
+				cgi.DrawString(x, y + 6, name, CON_COLOR_DEFAULT);
+
+				x -= NOTIFICATION_PIC_HEIGHT + NOTIFICATION_PADDING_X;
+
+				cgi.DrawImage(x, y, ((vec_t) NOTIFICATION_PIC_HEIGHT) / HUD_PIC_HEIGHT,
+					cgi.LoadImage(Bg_GetModIcon(item->mod, item->mod &
MOD_FRIENDLY_FIRE), IT_PIC));

Ditto, see below.
------------------------------

In src/cgame/default/cg_hud.c
<https://github.com/jdolan/quetoo/pull/487#discussion_r120530444>:

> +		item = (bg_notification_item_t *) list->data;
+
+		x = cgi.context->width - 10;
+
+		switch (item->type) {
+			case NOTIFICATION_TYPE_OBITUARY: {
+				char *name = cgi.client->client_info[item->client_id_2].name;
+
+				x -= cgi.StringWidth(name) + NOTIFICATION_PADDING_X;
+
+				cgi.DrawString(x, y + 6, name, CON_COLOR_DEFAULT);
+
+				x -= NOTIFICATION_PIC_HEIGHT + NOTIFICATION_PADDING_X;
+
+				cgi.DrawImage(x, y, ((vec_t) NOTIFICATION_PIC_HEIGHT) / HUD_PIC_HEIGHT,
+					cgi.LoadImage(Bg_GetModIcon(item->mod, item->mod &
MOD_FRIENDLY_FIRE), IT_PIC));

Ditto, see below.
------------------------------

In src/cgame/default/cg_hud.c
<https://github.com/jdolan/quetoo/pull/487#discussion_r120530549>:

> +			break;
+		case NOTIFICATION_TYPE_OBITUARY_SELF:
+			item->mod = cgi.ReadByte();
+			item->client_id_1 = cgi.ReadByte();
+
+			item->when = 6000;
+
+			if (!cg_draw_notifications_disable_print->value) {
+				cgi.Print("^1%s ^7%s\n", Bg_GetModString(item->mod, false),
+					cgi.client->client_info[item->client_id_1].name);
+			}
+
+			break;
+		case NOTIFICATION_TYPE_OBITUARY_PIC: {
+			const char *p = cgi.ReadString();
+			strcpy(item->pic, p);

"pic" should be cgi.LoadImage'd. I'm not certain, but if the obituary
pictures are entirely constant (that is, the pics you send are *always*
consistent between game and cgame for obituaries), send the index to the
array and not the pic itself. Then we can reference a cgame-specific array
of the images that are precached in this file.
------------------------------

In src/cgame/default/cg_hud.c
<https://github.com/jdolan/quetoo/pull/487#discussion_r120530559>:

> +			strcpy(item->pic, p);
+			item->client_id_1 = cgi.ReadByte();
+			item->client_id_2 = cgi.ReadByte();
+
+			item->when = 6000;
+
+			if (!cg_draw_notifications_disable_print->value) {
+				cgi.Print("^7%s ^1[%s] ^7%s\n",
cgi.client->client_info[item->client_id_1].name,
+					item->pic,
+					cgi.client->client_info[item->client_id_2].name);
+			}
+		}
+			break;
+		case NOTIFICATION_TYPE_PLAYER_ACTION: {
+			const char *p = cgi.ReadString();
+			strcpy(item->pic, p);

Ditto, cgi.LoadImage here.
------------------------------

In src/game/default/bg_notification.h
<https://github.com/jdolan/quetoo/pull/487#discussion_r120530588>:

> +	NOTIFICATION_TYPE_PLAYER_ACTION, // Race times, capture times (PIC + CID + STRING)
+} bg_notification_type_t;
+
+/**
+ * @brief Notification feed
+ */
+typedef struct bg_notification_item_s {
+	bg_notification_type_t type;
+
+	char string_1[MAX_STRING_CHARS]; // STRING
+	char string_2[MAX_STRING_CHARS]; // STRING
+
+	uint16_t client_id_1; // CID
+	uint16_t client_id_2; // CID
+
+	char pic[MAX_QPATH]; // PIC

should be uint16_t for a pic index.
------------------------------

In src/game/default/g_types.h
<https://github.com/jdolan/quetoo/pull/487#discussion_r120530702>:

> +#define MOD_LAVA					17
+#define MOD_CRUSH					18
+#define MOD_TELEFRAG				19
+#define MOD_FALLING					20
+#define MOD_SUICIDE					21
+#define MOD_EXPLOSIVE				22
+#define MOD_TRIGGER_HURT			23
+#define MOD_HANDGRENADE				24
+#define MOD_HANDGRENADE_SPLASH		25
+#define MOD_HANDGRENADE_SUICIDE		26
+#define MOD_HANDGRENADE_KAMIKAZE	27
+#define MOD_FIREBALL				28
+#define MOD_HOOK					29
+#define MOD_ACT_OF_GOD				30
+#define MOD_FRIENDLY_FIRE			0x80
+

Why dis not enum
------------------------------

In src/cgame/default/cg_hud.c
<https://github.com/jdolan/quetoo/pull/487#discussion_r120530766>:

>
-		Cg_DrawVital(x, health, health_icon, HUD_HEALTH_MED, HUD_HEALTH_LOW);
+		Cg_DrawVital(x, true, health, health_icon, HUD_HEALTH_MED, HUD_HEALTH_LOW);
 	}

 	if (atoi(cgi.ConfigString(CS_GAMEPLAY)) != GAME_INSTAGIB) {

One thing I've been meaning to do but haven't gotten around to is finding a
common place in cgame to store the results of these cgi.ConfigString fetch
operations. Seems silly to parse this every single frame multiple times.
------------------------------

In src/cgame/default/cg_hud.c
<https://github.com/jdolan/quetoo/pull/487#discussion_r120530796>:

> @@ -1107,12 +1144,29 @@ static void Cg_DrawSelectWeapon(const player_state_t *ps) {
 	r_pixel_t ch;
 	cgi.BindFont("medium", NULL, &ch);

+	if (cg_select_weapon_fade->modified || cg_select_weapon_interval->modified) {
+		cg_select_weapon_fade->modified = false;
+
+		cg_select_weapon_fade->value = Clamp(cg_select_weapon_fade->value,
0.0, cg_select_weapon_interval->value);
+	}

Fade is a good idea for this, I need to test it to see how it feels though.
If I was to implement fade it would be an additional amount of seconds
added to the display interval probably.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/jdolan/quetoo/pull/487#pullrequestreview-42498186>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/ANbJuVSeEgmBAgSXYAdSwNYh7GwTsfwaks5sBiJYgaJpZM4NxjYT>
.

kaadmy commented on src/game/default/g_client_stats.c:
```diff
@@ -270,6 +270,12 @@ void G_ClientStats(g_entity_t *ent) {
+	// respawn time
+	client->ps.stats[STAT_RESPAWN] = 0;
+	if (g_level.time < client->locals.respawn_time) {
+		client->ps.stats[STAT_RESPAWN] = client->locals.respawn_time - g_level.time;
```
Off the top of my head, I don't think the client knows about server time.
The timer and match state (timeout, warmup, countdown) is sent as a config string to all clients.

kaadmy commented on src/cgame/default/cg_hud.c:
```diff
@@ -1107,12 +1144,29 @@ static void Cg_DrawSelectWeapon(const player_state_t *ps) {
+		cg_select_weapon_fade->modified = false;
+
+		cg_select_weapon_fade->value = Clamp(cg_select_weapon_fade->value, 0.0, cg_select_weapon_interval->value);
+	}
```
Yeah that would make more sense, but this was easier :P

kaadmy commented on src/cgame/default/cg_hud.c:
```diff
@@ -1175,6 +1233,231 @@ static void Cg_DrawTargetName(const player_state_t *ps) {
+
+	item->when = cgi.client->unclamped_time + (item->when * cg_draw_notifications_time->value);
+
+	notifications.items = g_slist_prepend(notifications.items, item);
```
Items are always added, they're just not rendered.
I should probably make them not add if there isn't room or you could get notifications popping back into the screen sometimes.

jdolan replied:
This looks good!

kaadmy commented:
I believe this is now complete enough to merge, minus the g_slist stuff?

jdolan commented on src/cgame/default/cg_hud.c:
```diff
@@ -1175,6 +1233,231 @@ static void Cg_DrawTargetName(const player_state_t *ps) {
+		list = g_slist_nth(notifications.items, i);
+
+		if (list && cgi.client->unclamped_time > ((bg_notification_item_t *) list->data)->when) {
+			notifications.items = g_slist_delete_link(notifications.items, list);
```
Leaky?

jdolan commented on src/game/default/bg_notification.c:
```diff
@@ -0,0 +1,126 @@
+#include "bg_notification.h"
+
+/**
+ * @brief List of MOD strings
```
A Doxygen @see tag (or something) here pointing to the MOD defines might be nice.

jdolan commented on src/game/default/g_client.c:
```diff
@@ -199,7 +86,7 @@ static void G_ClientObituary(g_entity_t *self, g_entity_t *attacker, uint32_t mo
 			        attacker->client->locals.persistent.team) {
 
-				if (friendy_fire) {
+				if (friendly_fire) {
```
lol.

jdolan commented:
Looks like this PR is starting to rot a bit. Is it something you want to save @kaadmy ?

jdolan commented:
As long as we can pin down the memory leaks, and this doesn't introduce crashes or major bugs, I'm okay with merging it in rough state. The menus are rough right now anyway ;)

jdolan commented:
@kaadmy Do you have any interest in pursuing this branch further? If you do, I'm happy to resolve the conflicts here and put it back in a good state for you to continue.

jdolan commented on src/game/default/g_client_stats.c:
```diff
@@ -270,6 +270,12 @@ void G_ClientStats(g_entity_t *ent) {
+	// respawn time
+	client->ps.stats[STAT_RESPAWN] = 0;
+	if (g_level.time < client->locals.respawn_time) {
+		client->ps.stats[STAT_RESPAWN] = client->locals.respawn_time - g_level.time;
```
The client absolutely knows about server time. Every frame it receives includes an implicit timestamp. Agreed that re-sending stats every frame is a bit wasteful. As a compromise / means to an end, would it be adequate to send it once per second here, or something?

kaadmy replied:
It's for the respawn timer so it has to be at least every 0.1 seconds.

jdolan replied:
We can send this once, when the player dies, and the client can cache it in cg_state_t and run the countdown locally.

jdolan commented on src/cgame/default/cg_hud.c:
```diff
@@ -1175,6 +1233,231 @@ static void Cg_DrawTargetName(const player_state_t *ps) {
+		list = g_slist_nth(notifications.items, i);
+
+		if (list && cgi.client->unclamped_time > ((bg_notification_item_t *) list->data)->when) {
+			notifications.items = g_slist_delete_link(notifications.items, list);
```
Right, it looks like this leak is still here.

kaadmy replied:
Yeah there are obvious problems, for example the second time you load a map it crashes :3

jdolan commented:
This is now up to date against develop and compiles and runs. There are some bugs, but overall, this is a lot of progress in the right direction. It also makes inroads towards #546. @kaadmy do you have any interest in completing this work?

jdolan commented on src/game/default/bg_notification.c:
```diff
@@ -0,0 +1,128 @@
+	"pics/notifications/i_lava",
+	"pics/notifications/w_hook",
+	"pics/notifications/i_death"
+};
```
Would an array of structs be better for this, rather than co-indexed arrays?

kaadmy replied:
Probably. I just kept this format because it was used elsewhere in the code.

jdolan commented on src/game/default/g_item.c:
```diff
@@ -518,8 +566,15 @@ static _Bool G_PickupFlag(g_entity_t *ent, g_entity_t *other) {
-			                  other->client->locals.persistent.net_name, t->name);
+			bg_notification_item_t notification_item = {
+				.type = NOTIFICATION_TYPE_PLAYER_ACTION,
+				.pic = gi.ImageIndex(va("pics/notifications/i_flag%d_return", t->id + 1)),
```
Might be nice to precache these pics in the team structures at level load.

jdolan commented:
I've cherry picked everything from this PR except for the notifications feed. We'll be adding emoji support to all strings, which will use the notification icons and others. Thank you for all of this work @kaadmy! :)
